### PR TITLE
Add literal in interpolation rule

### PR DESCRIPTION
**Pull request by jacknoble**
Hopefully this is what you're looking for in https://github.com/lpil/dogma/issues/46 . Let me know if there are any test case you'd like me to add or changes you'd like made!

### CODE REVIEW

**Comment by lpil** [comment]
Hello! Thank you for contributing. Always great to get a new person involved :)

File: lib/dogma/rules/literal_in_interpolation.ex [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+  alias Dogma.Script
+  alias Dogma.Error
+
+  import Dogma.Util.ASTNode, only: [literal?: 1]
+
+  def test(script, _config = [] \\ []) do
+    script |> Script.walk( &check_node(&1, &2) )
+  end
+
+  defp check_node({:::, meta, args} = node, errors) do
```
Comment (lpil) [inline lib/dogma/rules/literal_in_interpolation.ex]: `#{}` is a call to `:::`? How strange, wouldn't have guessed that.

File: lib/dogma/rules/literal_in_interpolation.ex [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+       _ -> {node, errors}
+    end
+  end
+
+  defp check_node(node, errors) do
+    {node, errors}
+  end
+
+  defp get_interpolated_token(args) do
+    case args do
```
Comment (lpil) [inline lib/dogma/rules/literal_in_interpolation.ex]: Why do we need this error handling?

File: lib/dogma/util/ast_node.ex [hunk 3]
```diff
@@ -0,0 +1,32 @@
+defmodule Dogma.Util.ASTNode do
```
Comment (lpil) [inline lib/dogma/util/ast_node.ex]: Good idea :)

File: test/dogma/documentation/rules_list_test.exs [hunk 4]
```diff
@@ -9,7 +9,7 @@ defmodule Dogma.Documentation.RuleListTest do
     assert actual == expected, """
     docs/rules.md is out of date.
 
-    Run `mix run script/generate_documentation.ex` in order to regenerate
+    Run `mix run scripts/generate_documentation.exs` in order to regenerate
```
Comment (lpil) [inline test/dogma/documentation/rules_list_test.exs]: Oops, my bad.

File: test/dogma/util/ast_node_test.exs [hunk 5]
```diff
@@ -0,0 +1,5 @@
+defmodule Dogma.Util.ASTNodeTest do
+  use ShouldI
+
+  doctest Dogma.Util.ASTNode
```
Comment (lpil) [inline test/dogma/util/ast_node_test.exs]: Probably should have more tests here.

File: lib/dogma/rules/literal_in_interpolation.ex (hunk 1 above)
Comment (jacknoble) [inline lib/dogma/rules/literal_in_interpolation.ex]: So one pattern Jose matches with is [here](https://github.com/elixir-lang/elixir/blob/aa901b1f5ad0749fecaf1fe5d89bbabcabf62399/lib/elixir/lib/macro.ex#L632). So `:<<>> ` is the salient function and arguments to that function are string literals and interpolations whose pattern is 
```
{:::, _,
   [{{:., _, [Kernel, :to_string]}, [], [(thing getting interpolated)]}, {:binary, _, _}]}}
```

`:::` by itself definitely over matches which is the reason for the error handling below. I originally had the whole pattern in the function def but that's pretty long and unwieldy. I'm sure you have more experience figuring out what to match on, open to suggestions.

File: test/dogma/util/ast_node_test.exs (hunk 5 above)
Comment (jacknoble) [inline test/dogma/util/ast_node_test.exs]: Agreed. I'll try to add some later this evening.

File: lib/dogma/rules/literal_in_interpolation.ex (hunk 1 above)
Comment (lpil) [inline lib/dogma/rules/literal_in_interpolation.ex]: Lets go with this, we can refine it later if we discover problems.

**Comment by jacknoble** [comment]
Added some tests for the ASTNode module. Rewrote some of the the `literal?` function so it makes more sense for AST nodes.

File: lib/dogma/util/ast_node.ex [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+  def literal?(x) when is_binary  x do true end
+  def literal?(x) when is_boolean x do true end
+  def literal?(x) when is_list    x do true end
+  def literal?(x) when is_number  x do true end
+  def literal?({_,_})               do true end
+  def literal?({:{}, _, _})         do true end
+  def literal?({:%{}, _, _})        do true end
+  def literal?({:sigil_s,_,_})      do true end
+  def literal?({:sigil_S,_,_})      do true end
+  def literal?({:sigil_w,_,_})      do true end
```
Comment (lpil) [inline lib/dogma/util/ast_node.ex]: What about other sigils?

File: test/dogma/util/ast_node_test.exs [hunk 7]
```diff
@@ -0,0 +1,71 @@
+defmodule Dogma.Util.ASTNodeTest do
```
Comment (lpil) [inline test/dogma/util/ast_node_test.exs]: Looks good :)

File: lib/dogma/util/ast_node.ex (hunk 6 above)
Comment (jacknoble) [inline lib/dogma/util/ast_node.ex]: Updated to match all of them. I agree.

**Comment by jacknoble** [comment]
Rebased onto current master.

File: lib/dogma/util/ast_node.ex [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+  def literal?(x) when is_list    x do true end
+  def literal?(x) when is_number  x do true end
+  def literal?({_,_})               do true end
+  def literal?({:{}, _, _})         do true end
+  def literal?({:%{}, _, _})        do true end
+  def literal?(node) do
+    sigil?(node)
+  end
+
+  def sigil?({fun, _, _}) when is_atom fun do
```
Comment (lpil) [inline lib/dogma/util/ast_node.ex]: If we build functions for each of these by iterating over the letters of the alphabet we will check all sigils and have no runtime penalty. Additionally, this matches function calls like this:

```elixir
def sigil_transformer(x) do
  # ...
end
```

https://github.com/lpil/dogma/blob/0d0ece43081c28059e487bd0b681068c9d3ceb4a/lib/dogma/rules/quotes_in_string.ex#L23-L26

Comment (jacknoble) [inline lib/dogma/util/ast_node.ex]: Ooh, smart! Does it make sense to have macros for defining functions that match different classes of AST node? e.g.
```
def_for_sigils literal? ({sigil,_,_}) do false end
```
which would expand to define that function for all sigils?

Comment (lpil) [inline lib/dogma/util/ast_node.ex]: Sure, though maybe not for now. When we have another case in which we need all sigils we should probably make this. :)

**Pull request merged**